File: src/bingo-eval.ts
/** 75-ball and 90-ball bingo card validation + pattern evaluation. */

export const BINGO_75_SHOE = 75;
export const BINGO_90_SHOE = 90;
export const BINGO_75_CELLS = 25;
export const BINGO_90_CELLS = 27;
export const BINGO_CHUNK_BALLS = 5;

export const BingoPattern75 = {
  AnyLine: 0,
  FourCorners: 1,
  Blackout: 2,
} as const;
export type BingoPattern75 = (typeof BingoPattern75)[keyof typeof BingoPattern75];

const COL75_LO = [1, 16, 31, 46, 61];
const COL75_HI = [15, 30, 45, 60, 75];
const COL90_LO = [1, 10, 20, 30, 40, 50, 60, 70, 80];
const COL90_HI = [9, 19, 29, 39, 49, 59, 69, 79, 90];

export function packBingoTerm(won: boolean, completionIndex: number): number {
  return (won ? 1 : 0) * 256 + completionIndex;
}

export function unpackBingoTerm(packed: number): { won: boolean; completionIndex: number } {
  return { won: packed >= 256, completionIndex: packed % 256 };
}

function assertUniqueNonZero(cells: readonly number[]): void {
  const seen = new Set<number>();
  for (const cell of cells) {
    if (cell === 0) continue;
    if (seen.has(cell)) throw new Error(`duplicate bingo number ${cell}`);
    seen.add(cell);
  }
}

export function assertValidBingo75Card(cells: readonly number[]): void {
  if (cells.length !== BINGO_75_CELLS) {
    throw new Error(`75-ball card must have ${BINGO_75_CELLS} cells`);
  }
  if (cells[12] !== 0) throw new Error('75-ball center must be the free space (0)');
  for (let i = 0; i < BINGO_75_CELLS; i++) {
    const cell = cells[i]!;
    if (i === 12) continue;
    const col = i % 5;
    if (!Number.isInteger(cell) || cell < COL75_LO[col]! || cell > COL75_HI[col]!) {
      throw new Error(`75-ball cell ${i} must be in ${COL75_LO[col]}–${COL75_HI[col]}, got ${cell}`);
    }
  }
  assertUniqueNonZero(cells);
}

export function assertValidBingo90Card(cells: readonly number[]): void {
  if (cells.length !== BINGO_90_CELLS) {
    throw new Error(`90-ball ticket must have ${BINGO_90_CELLS} cells`);
  }
  for (let r = 0; r < 3; r++) {
    let filled = 0;
    for (let c = 0; c < 9; c++) {
      const cell = cells[r * 9 + c]!;
      if (cell === 0) continue;
      if (!Number.isInteger(cell) || cell < COL90_LO[c]! || cell > COL90_HI[c]!) {
        throw new Error(`90-ball cell ${r * 9 + c} must be in ${COL90_LO[c]}–${COL90_HI[c]} or 0, got ${cell}`);
      }
      filled += 1;
    }
    if (filled !== 5) throw new Error(`90-ball row ${r} must have 5 numbers, got ${filled}`);
  }
  assertUniqueNonZero(cells);
}

export function sampleBingo75Card(salt = 0): number[] {
  const cells = new Array<number>(25).fill(0);
  for (let col = 0; col < 5; col++) {
    const start = COL75_LO[col]!;
    for (let row = 0; row < 5; row++) {
      const i = row * 5 + col;
      if (i === 12) continue;
      cells[i] = start + ((row + salt) % 15);
    }
  }
  assertValidBingo75Card(cells);
  return cells;
}

export function sampleBingo90Card(salt = 0): number[] {
  const cells = new Array<number>(27).fill(0);
  for (let r = 0; r < 3; r++) {
    const cols = [0, 1, 3, 5, 8];
    for (let k = 0; k < 5; k++) {
      const c = cols[k]!;
      const lo = COL90_LO[c]!;
      cells[r * 9 + c] = lo + ((r + k + salt) % (COL90_HI[c]! - lo + 1));
    }
  }
  assertValidBingo90Card(cells);
  return cells;
}

function mark75(cells: readonly number[], called: readonly number[]): boolean[] {
  const marked = cells.map((cell, i) => i === 12 || cell === 0);
  const drawn = new Set(called.map((b) => b + 1));
  for (let i = 0; i < cells.length; i++) {
    if (cells[i] !== 0 && drawn.has(cells[i]!)) marked[i] = true;
  }
  return marked;
}

function anyLine75(marked: readonly boolean[]): boolean {
  for (let r = 0; r < 5; r++) {
    if ([0, 1, 2, 3, 4].every((c) => marked[r * 5 + c])) return true;
  }
  for (let c = 0; c < 5; c++) {
    if ([0, 1, 2, 3, 4].every((r) => marked[r * 5 + c])) return true;
  }
  if ([0, 6, 12, 18, 24].every((i) => marked[i])) return true;
  if ([4, 8, 12, 16, 20].every((i) => marked[i])) return true;
  return false;
}

function fourCorners75(marked: readonly boolean[]): boolean {
  return Boolean(marked[0] && marked[4] && marked[20] && marked[24]);
}

function blackout75(marked: readonly boolean[]): boolean {
  return marked.every(Boolean);
}

function patternHit75(marked: readonly boolean[], patternId: BingoPattern75): boolean {
  if (patternId === BingoPattern75.AnyLine) return anyLine75(marked);
  if (patternId === BingoPattern75.FourCorners) return fourCorners75(marked);
  return blackout75(marked);
}

function firstCompletion(
  nCalled: number,
  hitAt: (calledPrefix: number[]) => boolean,
  balls: readonly number[],
): number {
  for (let t = 0; t < nCalled; t++) {
    if (hitAt(balls.slice(0, t + 1) as number[])) return t + 1;
  }
  return 0;
}

export function evaluateBingo75(
  cells: readonly number[],
  balls: readonly number[],
  nCalled: number,
  patternId: BingoPattern75,
): { won: boolean; completionIndex: number; packed: number } {
  assertValidBingo75Card(cells);
  if (nCalled < 1 || nCalled > balls.length) {
    throw new Error(`nCalled ${nCalled} out of range for ${balls.length} balls`);
  }
  const completionIndex = firstCompletion(
    nCalled,
    (prefix) => patternHit75(mark75(cells, prefix), patternId),
    balls,
  );
  const won = completionIndex > 0;
  return { won, completionIndex, packed: packBingoTerm(won, completionIndex) };
}

function mark90(cells: readonly number[], called: readonly number[]): boolean[] {
  const drawn = new Set(called.map((b) => b + 1));
  return cells.map((cell) => cell !== 0 && drawn.has(cell));
}

function completedRows90(cells: readonly number[], marked: readonly boolean[]): number {
  let rows = 0;
  for (let r = 0; r < 3; r++) {
    let need = 0;
    let got = 0;
    for (let c = 0; c < 9; c++) {
      const i = r * 9 + c;
      if (cells[i] !== 0) {
        need += 1;
        if (marked[i]) got += 1;
      }
    }
    if (need === 5 && got === 5) rows += 1;
  }
  return rows;
}

export function evaluateBingo90(
  cells: readonly number[],
  balls: readonly number[],
  nCalled: number,
): {
  oneLine: { won: boolean; completionIndex: number; packed: number };
  twoLines: { won: boolean; completionIndex: number; packed: number };
  fullHouse: { won: boolean; completionIndex: number; packed: number };
} {
  assertValidBingo90Card(cells);
  if (nCalled < 1 || nCalled > balls.length) {
    throw new Error(`nCalled ${nCalled} out of range for ${balls.length} balls`);
  }
  const term = (need: number) => {
    const completionIndex = firstCompletion(
      nCalled,
      (prefix) => completedRows90(cells, mark90(cells, prefix)) >= need,
      balls,
    );
    const won = completionIndex > 0;
    return { won, completionIndex, packed: packBingoTerm(won, completionIndex) };
  };
  return { oneLine: term(1), twoLines: term(2), fullHouse: term(3) };
}

File: libzkcasino/src/zig/player/bingo_eval.zig
pub const BINGO_75_SHOE: usize = 75;
pub const BINGO_90_SHOE: usize = 90;
pub const BINGO_75_CELLS: usize = 25;
pub const BINGO_90_CELLS: usize = 27;

pub fn packTerm(won: bool, completion: u32) u32 {
    return (if (won) @as(u32, 1) else 0) * 256 + completion;
}

fn mark75(cells: []const u8, called: []const u8, marked: *[25]bool) void {
    @memset(marked, false);
    for (cells, 0..) |cell, i| {
        if (i == 12 or cell == 0) marked[i] = true;
    }
    var drawn = [_]bool{false} ** 76;
    for (called) |b| drawn[b + 1] = true;
    for (cells, 0..) |cell, i| {
        if (cell != 0 and drawn[cell]) marked[i] = true;
    }
}

fn anyLine75(m: *const [25]bool) bool {
    var r: usize = 0;
    while (r < 5) : (r += 1) {
        if (m[r * 5] and m[r * 5 + 1] and m[r * 5 + 2] and m[r * 5 + 3] and m[r * 5 + 4]) return true;
    }
    var c: usize = 0;
    while (c < 5) : (c += 1) {
        if (m[c] and m[5 + c] and m[10 + c] and m[15 + c] and m[20 + c]) return true;
    }
    if (m[0] and m[6] and m[12] and m[18] and m[24]) return true;
    if (m[4] and m[8] and m[12] and m[16] and m[20]) return true;
    return false;
}

fn patternHit75(m: *const [25]bool, pattern_id: u32) bool {
    if (pattern_id == 1) return m[0] and m[4] and m[20] and m[24];
    if (pattern_id == 2) {
        for (m) |v| {
            if (!v) return false;
        }
        return true;
    }
    return anyLine75(m);
}

pub fn evaluate75(cells: []const u8, balls: []const u8, n_called: usize, pattern_id: u32) struct { won: bool, completion: u32, packed_term: u32 } {
    var t: usize = 0;
    while (t < n_called) : (t += 1) {
        var marked: [25]bool = undefined;
        mark75(cells, balls[0 .. t + 1], &marked);
        if (patternHit75(&marked, pattern_id)) {
            const completion: u32 = @intCast(t + 1);
            return .{ .won = true, .completion = completion, .packed_term = packTerm(true, completion) };
        }
    }
    return .{ .won = false, .completion = 0, .packed_term = packTerm(false, 0) };
}

fn mark90(cells: []const u8, called: []const u8, marked: *[27]bool) void {
    var drawn = [_]bool{false} ** 91;
    for (called) |b| drawn[b + 1] = true;
    for (cells, 0..) |cell, i| {
        marked[i] = cell != 0 and drawn[cell];
    }
}

fn completedRows90(cells: []const u8, marked: *const [27]bool) u32 {
    var rows: u32 = 0;
    var r: usize = 0;
    while (r < 3) : (r += 1) {
        var need: u32 = 0;
        var got: u32 = 0;
        var c: usize = 0;
        while (c < 9) : (c += 1) {
            const i = r * 9 + c;
            if (cells[i] != 0) {
                need += 1;
                if (marked[i]) got += 1;
            }
        }
        if (need == 5 and got == 5) rows += 1;
    }
    return rows;
}

pub const Bingo90 = struct {
    one_line: u32,
    two_lines: u32,
    full_house: u32,
};

pub fn evaluate90(cells: []const u8, balls: []const u8, n_called: usize) Bingo90 {
    const term = struct {
        fn run(need: u32, cells_: []const u8, balls_: []const u8, n_called_: usize) u32 {
            var t: usize = 0;
            while (t < n_called_) : (t += 1) {
                var marked: [27]bool = undefined;
                mark90(cells_, balls_[0 .. t + 1], &marked);
                if (completedRows90(cells_, &marked) >= need) return packTerm(true, @intCast(t + 1));
            }
            return packTerm(false, 0);
        }
    };
    return .{
        .one_line = term.run(1, cells, balls, n_called),
        .two_lines = term.run(2, cells, balls, n_called),
        .full_house = term.run(3, cells, balls, n_called),
    };
}

pub fn sample75(salt: u8, out: *[25]u8) void {
    const lo = [_]u8{ 1, 16, 31, 46, 61 };
    @memset(out, 0);
    var col: u8 = 0;
    while (col < 5) : (col += 1) {
        var row: u8 = 0;
        while (row < 5) : (row += 1) {
            const i = row * 5 + col;
            if (i == 12) continue;
            out[i] = lo[col] + ((row + salt) % 15);
        }
    }
}
